### PR TITLE
chore!: refactored the `getTransactionCost` method

### PR DESCRIPTION
**Pull request by petertonysmith94**
- Closes: #2362

# Release notes

In this release, we:

- Refactored functionality of `Provider.getTransactionCost` to `Account.getTransactionCost`
- Deprecated the `Provider.getResourcesForTransaction` method

# Summary

- Refactored functionality for `Provider.getTransactionCost` to `Account.getTransactionCost`:
  - Separate concerns of estimating and fake funding a transaction.
  - Reducing unnecessary parameters, leading to a cleaner API.
- Deprecated the `Provider.getResourcesForTransaction` method:
  -  Lack of use-cases and deemed not required.

# Breaking Changes

- Refactored functionality for `Provider.getTransactionCost` to `Account.getTransactionCost` **and** changed estimation parameter from `quantitiesToContract` to `quantities`.

```ts
// Before
const provider = Provider.create(...);
const account = Wallet.generate({ ... }) || new Predicate(...);
const quantities: Array<CoinQuantityLike> = [
  { amount: 1000, assetId: provider.getBaseAssetId() }
];

provider.getTransactionCost(txRequest, {
  resourceOwner: account,
  quantitiesToContract: quantities,
})
```

```ts
// After
const provider = Provider.create(...);
const account = Wallet.generate({ ... }) || new Predicate(...);
const quantities: Array<CoinQuantityLike> = [
  { amount: 1000, assetId: provider.getBaseAssetId() }
];

account.getTransactionCost(txRequest, {
  quantities
});
```

- Deprecation of the `Provider.getResourcesForTransaction` method.

# Checklist

- [x] I **_added_** — `tests` to prove my changes
- [x] I **_updated_** — all the necessary `docs`
- [x] I **_reviewed_** — the entire PR myself, using the GitHub UI
- [x] I **_described_** — all breaking changes and the Migration Guide

### CODE REVIEW

File: packages/account/src/account.ts
Comment (Torres-ssf) [inline packages/account/src/account.ts:607]: By funding the TX with `Account.generateFakeResources` instead of `txRequestClone.fundWithFakeUtxos`, we could remove this check enterely.

File: packages/account/src/account.ts
Comment (Torres-ssf) [inline packages/account/src/account.ts:631]: What is the motivation for moving this method to the Account class?

File: packages/program/src/functions/base-invocation-scope.ts
Comment (Torres-ssf) [inline packages/program/src/functions/base-invocation-scope.ts:262]: If the `Account.getTransactionCost` is already doing the same thing we have here, why do we need this code here?

**Review by Torres-ssf — COMMENTED** [review]
Hey @petertonysmith94 I understand that this is a draft yet, just leave a couple of questions here to better understand your motivations

File: packages/account/src/account.ts
Comment (Torres-ssf) [reply packages/account/src/account.ts:607]: @petertonysmith94 what was the error?

File: packages/account/src/account.ts
Comment (Torres-ssf) [reply packages/account/src/account.ts:631]: @petertonysmith94 I was just curious about this. I am under the impression that we can remove `getResourcesForTransaction` entirely.

File: packages/program/src/functions/base-invocation-scope.ts
Comment (Torres-ssf) [reply packages/program/src/functions/base-invocation-scope.ts:262]: The `getTransactionCost` here can be called by any of the `dryRun` methods so if the account is not defined, we can simply generate a new one using the provider. This method is only used for estimation purposes.

File: packages/account/src/account.ts
Comment (petertonysmith94) [reply packages/account/src/account.ts:607]: @Torres-ssf encountered a couple issues:

- `fundWithFakeUtxos` adds an input for the `baseAssetId` (if no present), or updates the existing input [quantity](https://github.com/FuelLabs/fuels-ts/blob/3b27bac26463b39b8201db96147ec4ef091d3de0/packages/account/src/providers/transaction-request/transaction-request.ts#L596) for the estimate. I believe this is be why I'm experiencing the `InsufficientInputAmount` related errors.

- I also attempted to use the `generateFakeResources` alongside the `fundWithFakeUtxos` and removing the `populateTransactionPredicateData`. However, also encountered some issues with the correct `generateFakeResources` not being used via inheritance.

Might be beneficial for us to have a chat over huddle - as I'd like to get these breaking changes over the line ASAP.

Comment (Torres-ssf) [reply packages/account/src/account.ts:607]: @petertonysmith94 Sure, I've been trying to mimic this issue without success.

Maybe it will be better to discuss it over a huddle

Comment (Torres-ssf) [reply packages/account/src/account.ts:607]: @petertonysmith94 Sorry for the delay here but it took me a while to figure out what the error was.

The problem is that you forget to add an amount to fund the TX using the base asset, to cover the TX fee for the dry-run

```ts
  async getTransactionCost(
    transactionRequestLike: TransactionRequestLike,
    { signatureCallback, quantitiesToContract = [] }: TransactionCostParams = {}
  ): Promise<TransactionCost> {
    const txRequestClone = clone(transactionRequestify(transactionRequestLike));
    const baseAssetId = this.provider.getBaseAssetId();

    // Fund with fake UTXOs to avoid not enough funds error
    // Getting coin quantities from amounts being transferred
    const coinOutputsQuantities = txRequestClone.getCoinOutputsQuantities();
    // Combining coin quantities from amounts being transferred and forwarding to contracts
    const requiredQuantities = mergeQuantities(coinOutputsQuantities, quantitiesToContract);
    const resources = this.generateFakeResources(
      mergeQuantities(
        [...requiredQuantities],
        [{ assetId: baseAssetId, amount: bn('100000000000000000') }]
      )
    );

    txRequestClone.addResources(resources);

    const txCost = await this.provider.getTransactionCost(txRequestClone, {
      signatureCallback,
    });

    return {
      ...txCost,
      requiredQuantities,
    };
  }
```

Comment (petertonysmith94) [reply packages/account/src/account.ts]: 33d9b1b346cae412d901fda51a66451755477f1e

File: packages/account/src/account.ts
Comment (Torres-ssf) [inline packages/account/src/account.ts:522]: @petertonysmith94 Maybe we should make it clear here that this quantity in `baseAssetId` is being added to pay for the TX fee at the `dryRun`

File: packages/account/src/account.ts
Comment (Torres-ssf) [inline packages/account/src/account.ts:532]: @petertonysmith94 Perhaps this should be addressed on `Provider.getTransactionCost`

File: .changeset/short-sheep-divide.md
Comment (Torres-ssf) [inline .changeset/short-sheep-divide.md:5]: @petertonysmith94 Shouldn't be `minor` for everyone here?  

The packages `contract`, and `program` are using the `account` package, and `interfaces` is being used by `program`?

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
@petertonysmith94 Great job 👏

File: packages/account/src/account.ts
Comment (petertonysmith94) [reply packages/account/src/account.ts]: a14330cf21d98d3c9b5f0dab5f7914b0122fa704

File: packages/account/src/account.ts
Comment (petertonysmith94) [reply packages/account/src/account.ts]: f6f66789fc41f704493de07a6f125d1de69f2f40

File: .changeset/short-sheep-divide.md
Comment (petertonysmith94) [reply .changeset/short-sheep-divide.md]: The rationale behind this was that the `@fuel-ts/account` is the only package with breaking changes.

Changed: b9058e0812bb123d3f1dd1be1bae4eadbaddd9e8

**Comment by Torres-ssf** [comment]
@ petertonysmith94 We still have some tests failing here.

Please let me know if you are having a hard time addressing them.

File: packages/account/src/providers/provider.test.ts
Comment (maschad) [inline packages/account/src/providers/provider.test.ts:1076]: I think this test should still be run, otherwise if this test is being skipped what's the purpose of these changes?

File: packages/account/src/providers/provider.ts
Comment (maschad) [inline packages/account/src/providers/provider.ts:1127]: Given these changes are already breaking, why not just remove the `requiredQuantities` from the `TransactionCost` type?

File: packages/account/src/providers/provider.test.ts
Comment (petertonysmith94) [reply packages/account/src/providers/provider.test.ts:1076]: > TODO: validate if this test still makes sense

With the following todo - I am unsure about this test.

From what I can see, the `minFee` will never be `0` as we add `1` after the `calculateGasFee`:
https://github.com/FuelLabs/fuels-ts/blob/8776b79986b8c47b34a9954278c3d9eafd504c09/packages/account/src/providers/provider.ts#L1024-L1029

Do you think this test is still valid?

File: packages/account/src/providers/provider.ts
Comment (petertonysmith94) [reply packages/account/src/providers/provider.ts:1127]: The `requiredQuantities` is required for the `TransactionCost` result in the `Account`. I've chosen to share this type between the `Provider.getTransactionCost` and `Account.getTransactionCost` methods, hence the omission here.

File: packages/account/src/providers/provider.test.ts
Comment (maschad) [reply packages/account/src/providers/provider.test.ts:1076]: That's correct but I think the test ensures that code remains as such, so I think it's valid. But I don't have the full context.

**Review by arboleya — COMMENTED** [review]
Conflicts to fix.

File: packages/account/src/providers/provider.test.ts
Comment (Torres-ssf) [reply packages/account/src/providers/provider.test.ts:1076]: @petertonysmith94 This test can be skipped for now. We can unskip this after https://github.com/FuelLabs/fuels-ts/issues/2813 is solved.

If we cannot use a gas price of `0`, the `maxFee` will never be `0`.

> From what I can see, the minFee will never be 0 as we add 1 after the calculateGasFee:

@petertonysmith94 you are correct, since `calculateGasFee` is always adding `1`, even when the `gasPrice` is `0`, the `maxFee` | `minFee` will never be `0`. 

But this can serve as validation that we are always adding `1` to `maxFee` and `minFee`.

Comment (petertonysmith94) [reply packages/account/src/providers/provider.test.ts:1076]: @Torres-ssf I think I'll wait for the merge of #2814 - may as well resolve this as one. Unless you'd advise differently?

File: packages/account/src/account.ts
Comment (Torres-ssf) [inline packages/account/src/account.ts:506]: The property name `quantitiesToContract` could use a renaming as it is not always used as quantities being forwarded to contracts. 

The Account method [withdrawToBaseLayer](https://github.com/FuelLabs/fuels-ts/blob/f3453b9ea669575e3cf97ee50512bc9039845e0c/packages/account/src/account.ts#L464) uses the property to inform the amount of the base asset that should be withdrawn to the base layer.

This is not a hard requirement but in my opinion, since we are already refactoring and breaking things, we could rename this property. What do you think?

**Review by Torres-ssf — COMMENTED** [review]
@petertonysmith94 I feel like we need to update this doc [page](https://github.com/FuelLabs/fuels-ts/blob/master/apps/docs/src/guide/contracts/cost-estimation.md) to say that the `getTransactionCost` method is provided by the Account class instead of the Provider class, what do you think?

File: packages/account/src/providers/provider.test.ts
Comment (Torres-ssf) [reply packages/account/src/providers/provider.test.ts:1076]: @petertonysmith94 That is ok. We can wait for https://github.com/FuelLabs/fuels-ts/pull/2814

File: packages/account/src/account.ts
Comment (petertonysmith94) [reply packages/account/src/account.ts:506]: I agree, however, I wasn't sure of a suitable name. What about just `quantities`?

Comment (Torres-ssf) [reply packages/account/src/account.ts:506]: Sure, this one is good as any

File: packages/account/src/providers/provider.test.ts
Comment (petertonysmith94) [reply packages/account/src/providers/provider.test.ts:1076]: Will resolve this after this has been merged:
- https://github.com/FuelLabs/fuels-ts/pull/2814

File: packages/account/src/account.ts
Comment (petertonysmith94) [reply packages/account/src/account.ts]: - Fixed docs: 126107d7d6e69190c946a4ef20c2fc1b00537802
- Updated parameter: 4c81e4b4c0ebd11fddd1c3e3f326645b498534d9

**Review by Torres-ssf — APPROVED** [review]

**Review by maschad — APPROVED** [review]
Nice job!

**Review by arboleya — APPROVED** [review]

**Pull request merged**
